Code Review for Miss Bortoletti: App to retrieve location

francescobau opened the pull request:
### DISCLAIMER
- I've done all of this in 4 days.
- Since the Product Owner wants to see more testing, more UI and more code by me, the only way for me to do this in less than 4 days was to start over, with a new repository.
- There's a LOT of code to watch, but **Ms. Bortoletti can ignore all tests (LocationParserTest, GPSCommandHandlerTest and GPSCommandReceiverTest)**.
- There used to be more classes, but they were deleted, forcing the MainActivity to be more "heavy".

### Introduction
Hello! This is a simple app, with the scope to retrieve your lost phone's location.
So, if PhoneA is lost, and it has the app installed with permissions granted, we are still able to locate that phone with PhoneB, even if PhoneB doesn't have the app (so, even an iPhone or an old Nokia 3310 is able to retrieve your PhoneA's location!). This last thing is possible if PhoneB sends PhoneA an SMS message with a certain format, and it can obtain the location if and only if the PIN code sent to PhoneA is the same as PhoneA 's set PIN code.

### Description of the classes:
- GPSCommandReceiver: the background service, needed to see if an incoming message is a command or not.
- GPSCommandHandler: it's the core of the incoming and outgoing commands.
    - If GPSCommandReceiver's listener receives a valid (LOCATE) command, that comand is handled, and, after obtaining the location from method getCurrentLocation() (which is currently in MainActivity) , it sends it back to the one who requested the location.
    - If MainActivity wants to send a (LOCATE) command, it calls GPSCommandHandler to do it.
- LocationParser: it's an utility class, very useful to parse an android.Location instance, and to check if that instance is with default values or not.
- MainActivity: it literally does the rest. It has a static method getCurrentLocation(), needed to obtain the location, it communicates with the UI, so, if someone presses the button, it requests to send the (LOCATE) command, calling GPSCommandHandler's sendCommand(SMSMessage) method.

Have a good review!

### Known issues:
- Currently, PIN code is default, so basically everyone is able to find PhoneA 's location. This is a severe Security issue that needs to be fixed ASAP.
- The SMS library I use to send SMS isn't able to send messages with a "magic char", so it's impossible to send SMS with a URL, as an example.
- The SMS library version I use is 2.2, because the lastest version (3.0alpha) doesn't work with my code.
- App name is too generic.
- MainActivity is too "heavy".
- App crashes if someone presses the button with an empty field. (It's normal, because of the specifications, but I'll put a check in the near future).
- Inefficient way to check if an incoming message is a valid command or not.
- It works in foreground and active background only. It doesn't work if the app is fully-closed. (it might even crash multiple times, if only the service is active). The main reason is the static method getCurrentLocation() in MainActivity, which is very fragile, because it needs an Activity, but it crashes if it's null.
- Testing might not be completed.
- **Some testing methods fail because of MainActivity's `getCurrentLocation()` method**. It throws `NullPointerException`, because, currently, that method can't work if MainActivity is fully-closed. A Refactor is needed, but I can't do that now, because it's a drastic change. I'm gonna fix that in next versions.

### Future implementations:
- Background Service improvement -> **`GPSLocationManager` to retrieve current position.**
- Integration with Google Maps.
- RING command.
- PreActivity that shows the reason why we need certain permissions.
- Password (PIN code) management.


### Links
My repository: https://github.com/francescobau/AntiTheft
My branch: https://github.com/francescobau/AntiTheft/tree/feature-locationCommand-CR
SMSLibrary (version 2.2): https://github.com/Cogno-IDU/smslibrary/tree/v2.2

francescobau committed:
Files are ready to be Code-Reviewed.

francescobau committed:
Deleted Log.d(String,String), added a Toast, added new String in strings.xml.

francescobau committed:
Updated specifications in class GPSCommandReceiver, improved GPSCommandHandlerTest

KNOWN ISSUES:
- onCommandReceived failes, because the test force-closes the app, and getCurrentLocation() can't work if MainActivity is closed.

francescobau committed:
Improved testing in GPSCommandHandlerTest and in GPSCommandReceiverTest, added @ author and @ since in testing classes.

KNOWN ISSUES:
- Missing specifications in testing classes.
- onCommandReceived() test failes, because the test force-closes the app, and getCurrentLocation() can't work if MainActivity is closed (it needs an Activity instance to work).

francescobau committed:
Inserted warning in VALID_TELEPHONE_NUMBER, edited GPSCommandReceiverTest, added specifications in testing classes.

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/GPSCommandHandlerTest.java:
```diff
@@ -1,6 +1,103 @@
+     */
+    @Test(expected = NullPointerException.class)
+    public void sendNullMessage() {
+        new GPSCommandHandler().sendMessage(null);
```
You can use a private variable `GPSCommandHandler` rather than writing `new GPSCommandHandler` always.

`private GPSCommandHandler handler;`

in the ```
init(){
...
handler = new GPSCommandHandler();
}```

```suggestion
        handler.sendMessage(null);
```

francescobau replied:
You are totally right!
Though the test fails, because getCurrentLocation() can't work if the MainActivity is fully-closed, since it needs an Activity instance to retrieve the last location.

But thank you very much for the tip, I hadn't thought about that!

And don't be worried, every kind of suggestion is useful! ;)

francescobau replied:
You are totally right!
I don't know why I hadn't thought about that...
Maybe because I was in a rush....

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/GPSCommandReceiverTest.java:
```diff
@@ -1,5 +1,73 @@
+     */
+    @Test
+    public void onNullMessageReceived() {
+        new GPSCommandReceiver().onMessageReceived(null);
```
The same suggestion that I wrote in the GPSCommandHandlerTest.

```suggestion
        receiver.onMessageReceived(null);
```

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/LocationParserTest.java:
```diff
@@ -0,0 +1,179 @@
+     */
+    @Test
+    public void testLatitude() {
+        assertEquals(true, expectedLatitude == locationParser.getLatitude());
```
```suggestion
        assertEquals(expectedLatitude, locationParser.getLatitude());
```

Probably `locationParser.getLatitude()` returns an Object so you can't write `==` but you should write `.equals`.

You can apply the same advice to the methods below.

francescobau replied:
Actually no, because method `locationParser.getLatitude()` returns a double... `locationParser` is a `LocationParser` variable, and `LocationParser` class is made by me...

Here's a link to the class `LocationParser`:
https://github.com/francescobau/AntiTheft/blob/feature-locationCommand-CR/app/src/main/java/com/example/antitheft/structure/LocationParser.java

francescobau replied:
Also, assertEquals(double,double) is deprecated, that's why I was forced to do it this way...

bortoletti-giorgia replied:
You are right. You could write `assertTrue(expectedLatitude == locationParser.getLatitude());` or leave it as it is.

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/LocationParserTest.java:
```diff
@@ -0,0 +1,179 @@
+        // Supposing that we are actually there...
+        location.setAccuracy(NOT_DEFAULT_ACCURACY);
+        LocationParser parser = new LocationParser(location);
+        assertEquals(false, parser.isDefault());
```
```suggestion
        assertFalse(parser.isDefault());
```

francescobau replied:
Oh, I hadn't thought about that! Thank you for the tip!

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/LocationParserTest.java:
```diff
@@ -0,0 +1,179 @@
+    public void testDefault() {
+        Location location = new Location(LocationParser.DEFAULT_PROVIDER);
+        LocationParser parser = new LocationParser(location);
+        assertEquals(true, parser.isDefault());
```
```suggestion
        assertTrue(parser.isDefault());
```

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/LocationParserTest.java:
```diff
@@ -0,0 +1,179 @@
+        location.setLongitude(new Location(LocationParser.DEFAULT_PROVIDER).getLongitude());
+
+        LocationParser parser = new LocationParser(location);
+        assertEquals(true, parser.isDefault());
```
```suggestion
        assertTrue(parser.isDefault());
```

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/structure/LocationParser.java:
```diff
@@ -0,0 +1,137 @@
+public class LocationParser {
+    private static final String UNKNOWN_LOCATION_MESSAGE = "Location is NOT acquired.";
+    protected static final double DEFAULT_LOCATION = 0;
+    static final String DEFAULT_PROVIDER = "Default";
```
I suggest you to write always the type of variable (private / protected / public) to maintain a stylistic line.
But it isn't a mistake.

francescobau replied:
Good point here.
I'm gonna explain it right now...
- `DEFAULT_LOCATION` is supposed to be protected, because it might be useful to subclasses, as well. Though I'm gonna change it a bit, you made me realize something.
- `DEFAULT_PROVIDER` is a local thing. It was supposed to be used in `LocationManager` class as well, but I thought to put it package-private on purpose, so only the classes inside "structure" are able to access it. I've done this, because I wasn't thinking to make `DEFAULT_PROVIDER` accessible to subclasses as well...

The main question was: what if subclasses want to put a different `DEFAULT_PROVIDER` ?
That's why I decided this way...
But I'm not 100% sure if that's the right thing... What do you think about it?

francescobau replied:
In the end, I decided to delete `DEFAULT_LOCATION`.
But I still haven't edited `DEFAULT_PROVIDER`.
What do you think about it?

bortoletti-giorgia replied:
Maybe I explained myself wrong, I didn't mean to delete the variables but only to add private etc., for example
```suggestion
    protected static final String DEFAULT_PROVIDER = "Default";
```

If yu want to give the possibility to have several DEFAULT_PROVIDER, you could think of having one constant per class. Personally if it is DEFAULT it is the default for all and therefore I would make a single one.

francescobau replied:
No, I get what do you mean, and yes, I wanted it to be a local thing, accessible to the classes in package `structure` only, but, if you still think it's better to put it `protected` instead of `package-private`, well... I trust your opinion.
I'm gonna edit it, then....

And, I've deleted DEFAULT_LOCATION, because I could retrieve default Latitude and Longitude by doing `new Location().getLatitude()` and `new Location().getLongitude()` , so the constant was redundant.

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/structure/LocationParser.java:
```diff
@@ -0,0 +1,137 @@
+    static final String DEFAULT_PROVIDER = "Default";
+
+    private double latitude, longitude;
+    private boolean defaultFlag;
```
What I understand from `isDefault()` method is that `defaultFlag` indicates if a position has been acquired or not. Maybe I would call the variable differently, e.g. `acquiredLocation`. 
I tell you this because you do not say anywhere what this variable means. Otherwise I would tell you to explain in a comment what it is for.

francescobau replied:
You are right.
Though at first, I thought that way; in fact, it was called `isAcquired()` , and the field was called `acquired`, but then I also realized that even mocked Locations can have a (mocked) accuracy...
So "acquired" is not appropriate, otherwise, the check would have been more complex.

Also, at first, there was another constructor, which was `LocationParser(double,double)` , and method `setLocation(double,double)` was public. But then I decided to make it accessible with a `Location` value only, due to possible security issues, and because it started to become hard to check if it was actually acquired or not.

In the end, I decided to rename it `defaultFlag`, because it was easier to handle, so, if the `Location` is set with default values, `defaultFlag`  becomes true.

I'm gonna write more comments, though.
It seems like I hadn't clarified that enough.

Thank you!

bortoletti-giorgia replied:
Yes, exactly, try to give a more meaningful name or explain it better.

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/structure/GPSCommandReceiver.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    @Override
+    public void onMessageReceived(@Nullable SMSMessage message) {
+        if (message == null) return;
+        if (!isValidCommand(message)) return;
+        new GPSCommandHandler().onCommandReceived(message);
```
```suggestion
        if (message != null && isValidCommand(message))
                   new GPSCommandHandler().onCommandReceived(message);
```

In this way you do nothing if message isn't valid, without using `return;` in a void method.
Otherwise you can throw an exception if the format is not valid.
Or assemble:

```suggestion
        if (message == null || !isValidCommand(message))
            return;
        new GPSCommandHandler().onCommandReceived(message);
```

francescobau replied:
I prefer not to throw an exception, but just to ignore it, because that listener obtains **all** incoming messages.
Otherwise, if someone sends you a non-valid command via SMS, the app would crash.

Also, I've tried your second proposal, but, even though java is supposed to be "lazy", so if the message is null, then it's supposed to say _"ok, that condition is true, let's skip the other one!"_, strangely it doesn't do that for me... I tried that, but the test failed if I tried to give it a null `String` .

So that's the reason why I decided to split the conditions.

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/structure/GPSCommandHandler.java:
```diff
@@ -0,0 +1,66 @@
+    protected void sendLocation(@NonNull SMSPeer smsPeer, @Nullable String location) {
+        String text;
+        if (location == null) text = new LocationParser().toString();
+        else text = "Last known location: " + location;
```
```suggestion
        if (location == null)
           text = new LocationParser().toString();
        else
           text = "Last known location: " + location;
```

just a note, not a mistake

francescobau replied:
Well, thank you for the suggestion.
`CTRL+SHIFT+L` already said it's well-formatted, but I'm gonna try to apply it anyway.

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/structure/GPSCommandHandler.java:
```diff
@@ -0,0 +1,66 @@
+     * @see GPSCommandReceiver#onMessageReceived(SMSMessage)
+     * @see MainActivity#getCurrentLocation()
+     */
+    protected void onCommandReceived(@NonNull SMSMessage command) {
```
I would suggest you call the method differently because from the message is used only the peer and the command is the analyzed before calling this method. Like:
```suggestion
    protected void onCommandReceivedFrom(@NonNull SMSPeer peer) {
```

francescobau replied:
You are totally right! I hadn't thought about that...

Though that brings me to change a lot of things, by just changing it to SMSPeer...
I hope I won't forget something behind...

francescobau replied:
Hold on.
Actually, I need that to be called "`onCommandReceived()`", and I need a `SMSMessage` instance, because, in the future, there's not gonna be just the LOCATE command, but other commands as well, so, after `GPSCommandReceived` noticed that's a valid command, it leaves the rest to `GPSCommandHandler`.

But, you made me think I should better call this method `onLocateCommandReceived(@ NonNull SMSPeer peer)` instead, and, in the future, I'm gonna have more methods, e.g. `onRingCommandReceived(@ NonNull SMSPeer peer)`.

So basically, there are these 2 choices as well:
1) (my first idea) leaving the method as it is, and use a `switch-case` to check which kind of command is requested.
    - PRO: more reliable, if well-implemented, lighter GPSCommandReceiver.
    - CON: more complex, possible redundancy (GPSCommandReceiver already checks if it's a command)
2) rename it `onLocateCommandReceived(@ NonNull SMSPeer peer)` ,  and, in the future, add more similar methods.
    - PRO: easier method(s)
    - CON: GPSCommandReceiver needs to filter better the incoming messages, and the method is more fragile, if someone on purpose calls that method, even if it's not a command => double-check is needed.

Now that I think about it, I think (2) might be the best choice, but the issue is that `GPSCommandReceiver` needs to filter better, and we need to make these methods more reliable.
What do you think about it? One of these two, or the one you suggested?

bortoletti-giorgia replied:
In my opinion, you could keep in the GPSCommandReceiver the onMessageReceived method which analyzes the command received and invokes the right class.
In this GPSCommandHandler class you rename this method as onLocateCommandReceived. If in the future you want to add the ring option then you will do a RingCommandHandler and add there the method.

Then GPSCommandReceiver analyzes the command and invokes the right method either of GPSCommandHandler or of RingCommandHandler.

francescobau replied:
Ok, so you agree with option (2), too, right?

I'll change that when I can, then!

Thank you for your opinion! ✌️

francescobau replied:
Wait, what do you mean by "RingCommandHandler"?

You mean like proposal (2), so method `onRingCommandReceived(SMSPeer)` inside GPSCommandHandler (I guess I should find a better name, in the future...), or you mean another class?

bortoletti-giorgia replied:
Another class (e.g. `RingCommandHandler`) with method `onRingCommandReceived(SMSPeer)`.
Different managers/handlers for different scopes.
It is advice for keeping classes and uses separate.

francescobau replied:
Oh, I see....
Thank you for the advice! ✌️ 
I will surely think about that!

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/MainActivity.java:
```diff
@@ -0,0 +1,174 @@
+                boolean permissionsGranted = checkSelfPermission(Manifest.permission.SEND_SMS) == PERMISSION_GRANTED;
+                if (permissionsGranted) {
+                    String telephoneNumber = ((EditText) findViewById(R.id.telephoneNumber)).getText().toString();
+                    // Checking the peer.
```
To check the EditText you can add an if, like

```suggestion
                   if( !telephoneNumber.isEmpty()) {
                    // Checking the peer.
```

francescobau replied:
Yes, you are right!
I've done an edit like this:
```
if (telephoneNumber == null || telephoneNumber.isEmpty()) {
                        Toast.makeText(context, R.string.empty_telephone_number, Toast.LENGTH_SHORT).show();
                        return;
                    }
```
Is it ok? What do you think?

bortoletti-giorgia replied:
It's good for me.

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/MainActivity.java:
```diff
@@ -0,0 +1,174 @@
+                if (permissionsGranted) {
+                    String telephoneNumber = ((EditText) findViewById(R.id.telephoneNumber)).getText().toString();
+                    // Checking the peer.
+                    SMSPeer peer = new SMSPeer(telephoneNumber);
```
`new SMSPeer` throws `InvalidTelephoneNumberException` so you can use this exception to check the validity of the peer.

francescobau replied:
You are right!
I haven't thought about any kind of try-catch in my code... Especially because, in general, I have to be careful about using it, or I can ruin the variables' life cycle.

I'm gonna try to do it in next commit...
Thank you for the tip!

francescobau replied:
I think I improved it now... What do you think about it?

bortoletti-giorgia replied:
I think it's okay how you handled the exception.

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/MainActivity.java:
```diff
@@ -0,0 +1,174 @@
+    public static String getCurrentLocation() {
+        String locationTag = "LOCATION";
+        // Callback to retrieve the last known location.
+        client.getLastLocation().addOnSuccessListener(activity, new OnSuccessListener<Location>() {
```
You could remove the `private Activity activity` variable and use `this` directly above all because if you need a `MainActivity`'s method you don't have it in a normal `Activity` and then you can always use `this` to invoke the class itself.
Did you need that variable for anything else?

```suggestion
        client.getLastLocation().addOnSuccessListener(this, new OnSuccessListener<Location>() {
```

francescobau replied:
Well, that's the main issue...
I can't use `this` in a static method.
And, unfortunately, not even a Context is enough to add a listener.

But thank you for spotting that. I'm gonna add more `@ see` references in next commit.

francescobau replied:
I think I improved this now...
I also realized that Activity instance isn't mandatory, and it still works!

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/MainActivity.java:
```diff
@@ -0,0 +1,174 @@
+     * @see LocationParser#toString()
+     */
+    public static String getCurrentLocation() {
+        String locationTag = "LOCATION";
```
This could be a final variable.

francescobau replied:
Oh shoot, I forgot to delete it!
Thank you for spotting that.

Actually, it wasn't supposed to be final, because it was supposed to be used, for debug reason, as a tag in `Log.d(String tag, String message)` inside `getCurrentLocation()` only.
So, since there were 2 Logs, in order to give them a visible category, without hardwiring, I decided this way.

bortoletti-giorgia commented on app/src/main/java/com/example/antitheft/structure/GPSCommandReceiver.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    private boolean isValidCommand(@NonNull SMSMessage smsMessage) {
+        boolean flag = smsMessage.getData().contains(MainActivity.FULL_DEFAULT_LOCATE_COMMAND);
+        return flag;
```
```suggestion
        return smsMessage.getData().contains(MainActivity.FULL_DEFAULT_LOCATE_COMMAND);
```

francescobau replied:
I decided that way, because `smsMessage.getData()` can be null.
So I preferred to let it crash before it thinks to return an exception, when `smsMessage.getData()` is actually null.

francescobau replied:
I've just noted that `getData()` can't return a null instance. So I'm gonna edit it right away. Thank you for the suggestion.

bortoletti-giorgia reviewed (changes requested):
### Test
I have seen and ran also the tests because the code wasn't so much and I looked at them out of curiosity.
In GPSCommandHandlerTest e GPSCommandReceiverTest, tests fail and only methods of class to test, are invoked without any `Assert`.
You said that the tests were not finished so I have written only some things for the styling.

### In general
In general, some checks to avoid crashing the app with an _if statement_ could have been done before asking for revision and the tests are either complete or I personally wouldn't have delivered them. However I appreciated that you wrote the problems in the description of the pull request.
I recommend adding `getCurrentLocation()` to a hypothetical `LocationManager` and adding a Helper class to form the message to be sent and analyze the message received, so having a class that deals only with this and one that deals only with sending and receiving messages.
I really liked the specifics with @see, @version and also the @NonNull appropriate to the context.
I wrote several things but many are stylistic notes. I hope I haven't been too meticulous or severe. Tell me everything that doesn't fit you.

francescobau commented:
The test fails, because getCurrentLocation() can't work if the MainActivity is fully-closed, since it needs a non-null Activity instance to retrieve the last location.

Yeah, I was thinking about a LocationManager class as well. In fact, that's one of the classes I deleted, because I wasn't able to send an SMS with the acquired location.
After 2 days I realized it was because of a " **'** " inside the method sendLocation() in GPSCommandHandler, so, if location wasn't acquired, it worked, but, if location was acquired, it couldn't work!
It was like the app was trolling me, that was terrible...
But yeah, I'll definetly think about that.

And don't be worried, every kind of suggestion is useful to me! ;)

Thank you for the tips! ^^

francescobau committed:
Added private field in testing classes GPSCommandHandlerTest and GPSCommandReceiverTest.

Thank you Ms. Bortoletti for the tip.

francescobau committed:
Using Assert.assertTrue and Assert.assertFalse instead of Assert.assertEquals in LocationParserTest.

Thank you Ms. Bortoletti for the tips!

francescobau committed:
Deleted DEFAULT_LOCATION field from LocationParser, added explanation of defaultFlag field, general specifications improvement.

+ added strings in strings.xml
+ added specifications in GPSCommandReceiver
+ small reformat in GPSCommandHandler
+ fixed wrong specifications in LocationParser
+ improved specifications and comments in MainActivity
+ improved all if clauses in onClick() listener (MainActivity)

KNOWN ISSUES:
- onClick is too "heavy"

francescobau committed:
Deleted Activity instance in MainActivity. Improved specifications of method getCurrentLocation()

It's possible to add a onSuccessListener without an Activity instance, and it still works, so I removed it.

KNOWN ISSUES:
- service still crashes if a command is received at onDestroy() state (full-closed app).

francescobau commented:
Thank you very much for the tips!
I think I improved the code right now...
I also realized that the Activity instance isn't needed to retrieve the location, and it still works! ✌️ 🚀 
But I still have some doubts... If you can answer them in the unresolved conversations, that would be much appreciated! ✌️ 
Thank you in advance! ^^

francescobau commented:
Ah, I almost forgot...
`GPSCommandHandlerTest` and `GPSCommandReceivedTest` don't have any Assert, because they fail only if something unexpected happens (e.g. an Exception is thrown), since all methods are void, and only method `isValidCommand(@ NonNull SMSMessage)` isn't void... But it's a private method, so I can't test it.

If I had more time, I would have tried to mock these classes, but I don't know if it's possible to mock them...
At the beginning, I tried to mock `LocationParser`, but mockito told me that class isn't mockable, because they have private fields.

Any ideas?

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/GPSCommandHandlerTest.java:
```diff
@@ -1,6 +1,106 @@
+     * @see GPSCommandHandler#onCommandReceived(SMSMessage)
+     */
+    @Test
+    public void onCommandReceived() {
```
This method throws java.lang.NullPointerException for a null reference problem on the MainActivity.

francescobau replied:
Oops, my bad!
Thank you for spotting that...

francescobau replied:
> This method throws java.lang.NullPointerException for a null reference problem on the MainActivity.

I was wondering... Am I supposed to add a `@ throws` declaration in method `getCurrentLocation()` as well?

francescobau replied:
Wait... That's inside a Test class... I don't know if I'm supposed to do that, it should just fail when the unexpected happens...

francescobau replied:
It's not supposed to throw `NullPointerException`, but it does, because the method `getCurrentLocation` in `MainActivity` is not well-implemented. That method should be inside a Service, and I'm gonna do that in `develop` branch.
It's thanks to that test that I was able to know why does it not work at `onDestroy()` . So I can't remove it or "mask it" into something I expect. I'm gonna clarify it as a Known Issue, though... My apologies for not being clear enough.

bortoletti-giorgia commented on app/src/androidTest/java/com/example/antitheft/structure/GPSCommandReceiverTest.java:
```diff
@@ -1,5 +1,76 @@
+     * @see GPSCommandReceiver#onMessageReceived(SMSMessage)
+     */
+    @Test
+    public void onValidLocateCommandReceived() {
```
This method throws java.lang.NullPointerException for a null reference problem on the MainActivity.

francescobau replied:
Thank you. I completely forgot to add `@ throws` declarations!

francescobau replied:
Same thing as above, I'm not sure if I'm supposed to do that, since it's a Test class...

francescobau replied:
Same as above... It fails because method `getCurrentLocation()` is not well-implemented.
But I can't do drastic changes in this Pull Request, I'm gonna do it in `develop` branch, instead...

bortoletti-giorgia commented:
In my opinion LocationParserTest is okay and doesn't need to be mocked for now.
I wrote to you in the other two tests what exceptions it throws me and therefore you only need to manage those exceptions. Sorry for the delay.

francescobau committed:
Fixed and improved specifications, added @ version in testing classes, fixed some typos, added throws declaration in method getCurrentLocation

Thank you Miss Bortoletti for the tips!

francescobau commented:
My apologies for the delay.
I made some fixes and improvements in here...
I hope everything is OK.

For the `NullPointerException` issue, I edited the description, and I clarified it in the **Known Issues** section.

I've also added something, so I invite you to check the last commit I made...

francescobau committed:
Updated method isValidCommand(SMSMessage) in GPSCommandReceiver

Thank you Miss Bortoletti for the tip.
